src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs

```diff
@@ -0,0 +1,326 @@
+            }
+        }
+
+        /// <inheritdoc/>
+        public override int GetAvailablePermits() => _tokenCount;
+
+        /// <inheritdoc/>
+        protected override RateLimitLease AcquireCore(int tokenCount)
+        {
+            // These amounts of resources can never be acquired
+            if (tokenCount > _options.TokenLimit)
+            {
+                throw new ArgumentOutOfRangeException(nameof(tokenCount), $"{tokenCount} tokens exceeds the token limit of {_options.TokenLimit}.");
+            }
+
+            // Return SuccessfulLease or FailedLease depending to indicate limiter state
+            if (tokenCount == 0)
+            {
+                if (_tokenCount > 0)
+                {
+                    return SuccessfulLease;
+                }
+
+                return CreateFailedTokenLease(tokenCount);
+            }
+
+            lock (Lock)
+            {
+                if (TryLeaseUnsynchronized(tokenCount, out RateLimitLease? lease))
+                {
+                    return lease;
+                }
+
+                return CreateFailedTokenLease(tokenCount);
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override ValueTask<RateLimitLease> WaitAsyncCore(int tokenCount, CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // These amounts of resources can never be acquired
+            if (tokenCount > _options.TokenLimit)
+            {
+                throw new ArgumentOutOfRangeException(nameof(tokenCount), $"{tokenCount} token(s) exceeds the permit limit of {_options.TokenLimit}.");
+            }
+
+            // Return SuccessfulAcquisition if requestedCount is 0 and resources are available
+            if (tokenCount == 0 && _tokenCount > 0)
+            {
+                return new ValueTask<RateLimitLease>(SuccessfulLease);
+            }
+
+            lock (Lock)
+            {
+                if (TryLeaseUnsynchronized(tokenCount, out RateLimitLease? lease))
+                {
+                    return new ValueTask<RateLimitLease>(lease);
+                }
+
+                // Don't queue if queue limit reached
+                if (_queueCount + tokenCount > _options.QueueLimit)
+                {
+                    return new ValueTask<RateLimitLease>(CreateFailedTokenLease(tokenCount));
+                }
+
+                TaskCompletionSource<RateLimitLease> tcs = new TaskCompletionSource<RateLimitLease>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                CancellationTokenRegistration ctr = default;
+                if (cancellationToken.CanBeCanceled)
+                {
+                    ctr = cancellationToken.Register(obj =>
+                    {
+                        ((TaskCompletionSource<RateLimitLease>)obj!).TrySetException(new OperationCanceledException(cancellationToken));
```

Comment:
Added `!` to `obj`

Reply:
Same issue here with regards to a closure.